src/v2/params/types.ts

```diff
@@ -54,19 +71,41 @@ export class Param<T = unknown> {
     return this.toString();
   }
 
-  toSpec(): ParamSpec<string> {
+  toSpec(): ParamSpec<T> {
+    if (this.options.as) {
+      throw new Error(
+        `Param "${this.name}" has an "as" field, which is only valid for secret params.`
+      );
+    }
     const out: ParamSpec = {
       name: this.name,
       ...this.options,
       type: (this.constructor as typeof Param).type,
     };
-    if (this.options.default && typeof this.options.default !== 'string') {
-      out.default = (this.options.default as
-        | { toString?: () => string }
-        | undefined)?.toString?.();
-    }
 
-    return out as ParamSpec<string>;
+    return out as ParamSpec<T>;
+  }
+}
+
+export class SecretParam extends Param<string> {
```

Comment:
I personally wouldn't make SeecretParam inherit from Param<string> because Param<string> is an Expression<string> and SecretParam is not.